src/data/file.ts

```diff
@@ -479,6 +479,33 @@ export function parsePage(file: TFile, cache: MetadataCache, markdownData: Parse
             let frontFields = parseFrontmatter(fileCache.frontmatter) as Record<string, LiteralValue>;
             for (let [key, value] of Object.entries(frontFields)) fields.set(key, value);
         }
+
+        if (fileCache.headings && fileCache.headings.length > 0) {
+            let orderedLines: number[] = [];
+            let lineLookup = new Map<number, string>();
+            for (let heading of fileCache.headings) {
+                orderedLines.push(heading.position.start.line);
+                lineLookup.set(heading.position.start.line, heading.heading);
+            }
+
+            let nextHeader = orderedLines.shift();
+            let currentHeader = "";
+
+            function closestSection(line: number): string {
+                if (nextHeader && line >= nextHeader) {
+                    currentHeader = lineLookup.get(nextHeader) || "";
+                    nextHeader = orderedLines.shift();
+                }
+                return currentHeader;
+            }
+
+            markdownData.tasks.forEach(t => {
+                if (t.blockId) {
+                    return;
+                }
+                t.blockId = closestSection(t.line);
+            });
+        }
```

Comment:
In the medium term, this is likely better done inside the markdown parser itself than afterwards, though that will require refactors that aren't worth doing right now.

This code seems to be a lot for matching tasks to sections - why not something like

```
for (let task of markdownData.tasks) {
    if (task.blockId) continue;
    let nextSection = fileCache.headings.find(h => h.position.start.line > task.line);
    if (nextSection == 0) continue;

   task.blockId = fileCache.headings[nextSection - 1].heading;
}
```

Reply:
Agreed on the work being done in the parser. I'd been noodling about this process offline and I suspect that there's a straightforward way to not even depend upon the index, given that the hash is merely the content of a header. I'll simplify and see how much of a lift it is to get into the parser.

Reply:
I'm actually working on providing the full metadata cache to the parser (it's not actually that hard), which will make this code much easier for us. I can split that change off of a bunch of other junk I have and add it for you.

Reply:
Happen to have this handy? I can pull it in if you link me to a commit.